File: demo/fib_2min.go
package main

import (
	"fmt"
	"runtime"
	"time"
)

// seqFib 纯 CPU 密集型朴素递归计算（无缓存）
func seqFib(n int) int64 {
	if n <= 1 {
		return int64(n)
	}
	return seqFib(n-1) + seqFib(n-2)
}

// expandTree 将 Fibonacci 树展开指定深度，生成并发叶子任务列表
func expandTree(n int, depth int) []int {
	if depth <= 0 || n <= 1 {
		return []int{n}
	}
	left := expandTree(n-1, depth-1)
	right := expandTree(n-2, depth-1)
	return append(left, right...)
}

// parallelFib 基于 Go Channel 的树形分治并行算法
func parallelFib(n int, depth int) int64 {
	leaves := expandTree(n, depth)
	taskCount := len(leaves)
	ch := make(chan int64, taskCount)

	// 将各个叶子节点分发到独立 goroutine 并行计算
	for _, subN := range leaves {
		go func(sn int) {
			ch <- seqFib(sn)
		}(subN)
	}

	// 收集并累加所有叶子结果
	var sum int64 = 0
	for i := 0; i < taskCount; i++ {
		sum += <-ch
	}
	return sum
}

func main() {
	workers := runtime.NumCPU()

	fmt.Println("========================================================")
	fmt.Println("  Golang 原生 2分钟多核挑战：能算到 fib 几？")
	fmt.Println("  算法：基于 Go Channel 的树形分治并行 Fibonacci")
	fmt.Println("========================================================\n")

	fmt.Printf("检测到可用 CPU 工作核心数: %d\n", workers)

	const targetDuration = 120 * time.Second
	fmt.Printf("挑战目标时长: 120 秒 (2 分钟)\n")
	fmt.Println("并发机制：每阶 Fibonacci 树形展开为 32 个 Channel 并发叶子任务，各核心全负荷分治攻坚！")
	fmt.Println("请在另一个终端执行 htop 查看 CPU 占用（整整 2 分钟全部打满 2000%）！\n")

	startTime := time.Now()
	deadline := startTime.Add(targetDuration)

	// 从 35 阶开始冲刺（展开 5 层 = 32 个并发子任务）
	currentN := 35
	const depth = 5
	maxCompletedN := 0
	var maxCompletedVal int64 = 0

	fmt.Println("多核分治引擎已全线开火，开始冲击更高阶 Fibonacci...\n")

	for time.Now().Before(deadline) {
		t0 := time.Now()
		ans := parallelFib(currentN, depth)
		t1 := time.Now()

		cost := t1.Sub(t0).Seconds()
		totalElapsed := t1.Sub(startTime).Seconds()
		remSec := int(targetDuration.Seconds() - totalElapsed)
		if remSec < 0 {
			remSec = 0
		}

		maxCompletedN = currentN
		maxCompletedVal = ans

		fmt.Printf("  [剩余 %ds | 耗时 %.4fs] 攻克 fib(%d) = %d\n", remSec, cost, currentN, ans)

		currentN++
	}

	fmt.Println("\n2 分钟挑战时间到！正在生成统计战报...")

	totalSec := time.Since(startTime).Seconds()

	fmt.Println("\n========================================================")
	fmt.Println("  2 分钟多核极限冲刺完成！")
	fmt.Printf("实际总运行时间: %.2f 秒\n", totalSec)
	fmt.Printf("利用 %d 个核心，2 分钟内成功攻克的最大阶数是:\n\n", workers)
	fmt.Printf("    >>>  fib(%d) = %d  <<<\n\n", maxCompletedN, maxCompletedVal)
	fmt.Println("========================================================")
}

File: demo/fib_2min.py
#!/usr/bin/env python3
#
# Copyright (C) 2026 The Goldfish Scheme Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.
#

import os
import time
from multiprocessing import Pool, cpu_count

# seq_fib 纯 CPU 密集型朴素递归计算（无缓存）
def seq_fib(n: int) -> int:
    if n <= 1:
        return n
    return seq_fib(n - 1) + seq_fib(n - 2)

# expand_tree 将 Fibonacci 树展开指定深度，生成并发叶子任务列表
def expand_tree(n: int, depth: int) -> list:
    if depth <= 0 or n <= 1:
        return [n]
    return expand_tree(n - 1, depth - 1) + expand_tree(n - 2, depth - 1)

# parallel_fib 基于 Python 原生多进程池与队列的树形分治并行算法
def parallel_fib(n: int, depth: int, pool: Pool) -> int:
    leaves = expand_tree(n, depth)
    # 将 32 个子树叶子任务分发给进程池并行计算，收集结果并累加
    results = pool.map(seq_fib, leaves)
    return sum(results)

def main():
    workers = cpu_count()

    print("========================================================")
    print("  Python 原生多进程 2分钟多核挑战：能算到 fib 几？")
    print("  算法：基于 multiprocessing 树形分治并行 Fibonacci")
    print("========================================================\n")

    print(f"检测到可用 CPU 工作核心数: {workers}")

    target_duration_sec = 120
    print(f"挑战目标时长: {target_duration_sec} 秒 (2 分钟)")
    print("并发机制：利用 multiprocessing 绕过 GIL，每阶展开为 32 个叶子任务打满 20 个核心！")
    print("请在另一个终端执行 htop 查看 CPU 占用（整整 2 分钟全部打满 2000%）！\n")

    start_time = time.time()
    deadline = start_time + target_duration_sec

    # 从 35 阶开始冲刺（展开 5 层 = 32 个并发子任务）
    current_n = 35
    depth = 5
    max_completed_n = 0
    max_completed_val = 0

    print("多核分治引擎已全线开火，开始冲击更高阶 Fibonacci...\n")

    # 预创建常驻进程池，避免频繁创建销毁子进程
    with Pool(processes=workers) as pool:
        while time.time() < deadline:
            t0 = time.time()
            ans = parallel_fib(current_n, depth, pool)
            t1 = time.time()

            cost = t1 - t0
            total_elapsed = t1 - start_time
            rem_sec = int(max(0, target_duration_sec - total_elapsed))

            max_completed_n = current_n
            max_completed_val = ans

            print(f"  [剩余 {rem_sec}s | 耗时 {cost:.4f}s] 攻克 fib({current_n}) = {ans}")

            current_n += 1

    print("\n2 分钟挑战时间到！正在生成统计战报...")

    total_sec = time.time() - start_time

    print("\n========================================================")
    print("  2 分钟多核极限冲刺完成！")
    print(f"实际总运行时间: {total_sec:.2f} 秒")
    print(f"利用 {workers} 个核心，2 分钟内成功攻克的最大阶数是:\n")
    print(f"    >>>  fib({max_completed_n}) = {max_completed_val}  <<<\n")
    print("========================================================")

if __name__ == "__main__":
    main()
